### PR TITLE
Crimre 89 use client gem

### PR DESCRIPTION
**Pull request by willmcb**
## Description of change
Integrates review with the [types gem](https://github.com/ministryofjustice/laa-criminal-legal-aid-schemas), and the [datastore client api gem ](https://github.com/ministryofjustice/laa-criminal-applications-datastore-api-client)

## Link to relevant ticket
[CRIMRE-89](https://dsdmoj.atlassian.net/browse/CRIMRE-89)

## Notes for reviewer
- removes the old dev api code as well

## How to manually test the feature
- to test locally you will need to set up the data store and post some data to it. See the [datastore repo](https://github.com/ministryofjustice/laa-criminal-applications-datastore) to do this
- after this is set up test the app as if you are a user:
  - view page
  - click case
  - assign to self
  - un assign
  - etc...

### CODE REVIEW

File: app/views/crime_applications/_client_details.html.erb [hunk 1]
```diff
@@ -8,58 +8,58 @@
...
+      <%= client_details.applicant.telephone_number %>
     </dd>
   </div>
   <div class="govuk-summary-list__row">
     <dt class="govuk-summary-list__key">
       <%= t(:home_address, scope: 'crime_applications.labels') %>
     </dt>
     <dd class="govuk-summary-list__value">
-      <%= render client_details.client.address %>
+      <%= render 'crime_applications/address', address: client_details.applicant.address %>
```
Comment (zheileman) [inline app/views/crime_applications/_client_details.html.erb:51]: Beware in latest version of the schema this is named `home_address` 😬 :sorry

File: app/views/crime_applications/_address.html.erb [hunk 2]
```diff
@@ -3,7 +3,7 @@
 <br>
 <%= address.address_line_two %>
 <br>
-<%= address.county %>
+<%= address.country %>
```
Comment (zheileman) [inline app/views/crime_applications/_address.html.erb:6]: You also get a `city` attribute, in addition to `country`.

File: app/models/crime_application.rb [hunk 3]
```diff
@@ -1,19 +1,11 @@
...
-  attribute :case_details, CaseDetails
-  attribute :interests_of_justice, Types::Array.of(InterestOfJustice)
+class CrimeApplication < LaaCrimeSchemas::Structs::CrimeApplication
+  include ApiResource
 
   def applicant_name
-    client = client_details.client
-    [client.first_name, client.last_name].join ' '
+    applicant = client_details.applicant
+    [applicant.first_name, applicant.last_name].join ' '
```
Comment (zheileman) [inline app/models/crime_application.rb:8]: n a recent version of the schemas gem, there is a full_name method on any struct of type `Person` (so `Applicant` and `Codefendant` both have it). So you can do:

`client_details.applicant.full_name`

File: app/controllers/reassigns_controller.rb [hunk 4]
```diff
@@ -10,16 +10,25 @@ def new
   end
 
   def create
+    reassign_to_self
+  rescue Assigning::StateHasChanged
+    flash_and_redirect(:important, :state_has_changed)
+  else
```
Comment (timpeat) [inline app/controllers/reassigns_controller.rb:16]: 👍

File: app/views/assigned_applications/index.html.erb [hunk 5]
```diff
@@ -30,7 +30,9 @@
         </tr>
       </thead>
       <tbody class="govuk-table__body">
-        <%= render @applications %>
+        <% @applications.each do |app| %>
```
Comment (timpeat) [inline app/views/assigned_applications/index.html.erb:33]: Could remove the block and use collection here. E.g.

```suggestion
  <%= render 'application/crime_application', collection: @applications %>
```

File: app/views/crime_applications/_overview.html.erb [hunk 6]
```diff
@@ -46,7 +46,7 @@
       <%= t(:application_start_date, scope: 'crime_applications.labels') %>
     </dt>
     <dd class="govuk-summary-list__value">
-      <%= l(overview.application_start_date) %>
+      <%= l(overview.created_at) %>
```
Comment (timpeat) [inline app/views/crime_applications/_overview.html.erb:49]: I'm not mad about calling this created_at. Would prefer application_started_at. (Thinking about re-submissions etc). But that should not block this PR.

**Review by timpeat — APPROVED** [review]
So brilliant to get this in! Thank you! Once on staging we can get the DevAPI removed from apply etc.

**Pull request merged**